cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,973 @@
+The parent has two children. 
+
+1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each list
+   in a dense column of elements.
+2. A column containing the actual data and optional null mask for all elements of all the lists 
+   packed together.
+   
+With this representation, `data[offsets[i]]` is the first element of list `i`, and the size of list
+`i` is given by `offsets[i+1] - offsets[i]`.
+
+Note that the data may be of any type, and therefore the data column may itself be a nested column
+of any type. Note also that not only is each list nullable (using the null mask of the parent), but
+each list element may be nullable. So you may have a lists column with null row 3, and also null
+element 2 of row 4.
+
+## Strings columns
+
+Strings are represented in much the same way as lists, except that the data child column is always 
+a non-nullable column of `INT8` data. The parent column's type is `STRING` and contains no data,
+but its size represents the number of strings in the column, and its null mask represents the 
+validity of each string. To summarize, the strings column children are:
+
+1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each 
+   string in a dense column of all characters.
+2. A non-nullable column of `INT8` elements of all the characters across all the strings packed 
+   together.
+
+With this representation, `characters[offsets[i]]` is the first character of string `i`, and the 
+size of string `i` is given by `offsets[i+1] - offsets[i]`. The following image shows an example of
+this compound column representation of strings.
+
+![strings](strings.png)
+
+## Structs columns
+
+Structs are represented similarly to lists, except that they have multiple child data columns.
+The parent column's type is `STRUCT` and contains no data, but its size represents the number of 
+structs in the column, and its null mask represents the validity of each struct element. The parent 
+has `N + 1` children, where `N` is the number of fields in the struct. 
+
+1. A non-nullable column of `INT32` elements that indicates the offset to the beginning of each 
+   struct in each dense column of elements.
+2. For each field, a column containing the actual field data and optional null mask for all elements
+   of all the structs packed together.
+   
+With this representation, `child[0][offsets[i]]` is the first field of struct `i`, 
+`child[1][offsets[i]]` is the second field of struct `i`, etc.
+
+## Dictionary columns
+
+Dictionaries provide an efficient way to represent low-cardinality data by storing a single copy 
+of each value. A dictionary comprises a column of sorted keys and a column containing an index into the 
+keys column for each row of the parent column. The keys column may have any libcudf data type, 
+such as a numerical type or strings. The indices represent the corresponding positions of each 
+element's value in the keys.
```

Comment:
```suggestion
such as a numerical type or strings. The indices are represent the corresponding positions of each 
element's value in the keys. The indices child column can be any unsigned integer type (UINT8, UINT16, UINT32, or UINT64).
```

Reply:
I don't think the added `are` is correct.

Reply:
Added the sentence about index types.